File: app/src/main/java/com/xeniac/warrantyroster/landingactivity/loginfragment/LoginFragment.java
package com.xeniac.warrantyroster.landingactivity.loginfragment;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.Log;
import android.util.Patterns;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.res.ResourcesCompat;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;

import com.apollographql.apollo.ApolloCall;
import com.apollographql.apollo.ApolloClient;
import com.apollographql.apollo.api.Response;
import com.apollographql.apollo.exception.ApolloException;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.xeniac.warrantyroster.Constants;
import com.xeniac.warrantyroster.NetworkHelper;
import com.xeniac.warrantyroster.R;
import com.xeniac.warrantyroster.databinding.FragmentLoginBinding;
import com.xeniac.warrantyroster.mainactivity.MainActivity;

import java.util.Arrays;

public class LoginFragment extends Fragment {

    private FragmentLoginBinding loginBinding;
    private View view;
    private Activity activity;
    private Context context;
    private NavController navController;

    public LoginFragment() {
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        loginBinding = FragmentLoginBinding.inflate(inflater, container, false);
        view = loginBinding.getRoot();
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        loginBinding = null;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        activity = getActivity();
        context = getContext();
        navController = Navigation.findNavController(view);

        textInputsBackgroundColor();
        textInputsStrokeColor();
        forgotPwOnClick();
        registerOnClick();
        loginOnClick();
        loginActionDone();
    }

    private void textInputsBackgroundColor() {
        loginBinding.tiLoginEditEmail.setOnFocusChangeListener((view, focused) -> {
            if (focused) {
                loginBinding.tiLoginLayoutEmail.setBoxBackgroundColorResource(R.color.windowBG);
            } else {
                loginBinding.tiLoginLayoutEmail.setBoxBackgroundColorResource(R.color.grayLight);
            }
        });

        loginBinding.tiLoginEditPassword.setOnFocusChangeListener((view, focused) -> {
            if (focused) {
                loginBinding.tiLoginLayoutPassword.setBoxBackgroundColorResource(R.color.windowBG);
            } else {
                loginBinding.tiLoginLayoutPassword.setBoxBackgroundColorResource(R.color.grayLight);
            }
        });
    }

    private void textInputsStrokeColor() {
        loginBinding.tiLoginEditEmail.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence inputEmail, int start, int before, int count) {
                loginBinding.tiLoginLayoutEmail.setErrorEnabled(false);
                loginBinding.tiLoginLayoutEmail.setBoxStrokeColor(context.getResources().getColor(R.color.blue));
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        loginBinding.tiLoginEditPassword.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence inputEmail, int start, int before, int count) {
                loginBinding.tiLoginLayoutPassword.setErrorEnabled(false);
                loginBinding.tiLoginLayoutPassword.setBoxStrokeColor(context.getResources().getColor(R.color.blue));
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
    }

    private void forgotPwOnClick() {
        loginBinding.btnLoginForgotPw.setOnClickListener(view ->
                navController.navigate(R.id.action_loginFragment_to_forgotPasswordFragment));
    }

    private void registerOnClick() {
        loginBinding.btnLoginRegister.setOnClickListener(view ->
                navController.navigate(R.id.action_loginFragment_to_registerFragment));
    }

    private void loginOnClick() {
        loginBinding.btnLoginLogin.setOnClickListener(view ->
                loginViaEmail());
    }

    private void loginActionDone() {
        loginBinding.tiLoginEditPassword.setOnEditorActionListener((textView, actionId, keyEvent) -> {
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                loginViaEmail();
            }
            return false;
        });
    }

    private void loginViaEmail() {
        if (NetworkHelper.hasNetworkAccess(context)) {
            getLoginInputs();
        } else {
            hideLoadingAnimation();
            Snackbar snackbar = Snackbar.make(view,
                    context.getResources().getString(R.string.network_error_connection),
                    BaseTransientBottomBar.LENGTH_INDEFINITE)
                    .setAction(context.getResources().getString(R.string.network_error_retry), v -> loginViaEmail())
                    .setBackgroundTint(context.getResources().getColor(R.color.red))
                    .setTextColor(context.getResources().getColor(R.color.white))
                    .setActionTextColor(context.getResources().getColor(R.color.white))
                    .setAnimationMode(BaseTransientBottomBar.ANIMATION_MODE_SLIDE);
            TextView snackText = (snackbar.getView()).findViewById(R.id.snackbar_text);
            TextView snackAction = (snackbar.getView()).findViewById(R.id.snackbar_action);
            snackText.setTypeface(ResourcesCompat.getFont(context, R.font.nunito_semi_bold));
            snackAction.setTypeface(ResourcesCompat.getFont(context, R.font.nunito_bold));
            snackbar.show();
        }
    }

    private void getLoginInputs() {
        String email;
        String password;
        String deviceName = Build.MANUFACTURER + " " + Build.MODEL;

        if (TextUtils.isEmpty(loginBinding.tiLoginEditEmail.getText())) {
            loginBinding.tiLoginLayoutEmail.requestFocus();
            loginBinding.tiLoginLayoutEmail.setBoxStrokeColor(context.getResources().getColor(R.color.red));
        } else if (TextUtils.isEmpty(loginBinding.tiLoginEditPassword.getText())) {
            loginBinding.tiLoginLayoutPassword.requestFocus();
            loginBinding.tiLoginLayoutPassword.setBoxStrokeColor(context.getResources().getColor(R.color.red));
        } else {
            email = loginBinding.tiLoginEditEmail.getText().toString().trim().toLowerCase();
            password = loginBinding.tiLoginEditPassword.getText().toString().trim();

            if (!isEmailValid(email)) {
                loginBinding.tiLoginLayoutEmail.requestFocus();
                loginBinding.tiLoginLayoutEmail.setError(context.getResources().getString(R.string.login_error_email));
            } else {
                loginViaEmailMutation(email, password, deviceName);
            }
        }
    }

    private void loginViaEmailMutation(String email, String password, String deviceName) {
        showLoadingAnimation();

        ApolloClient apolloClient = ApolloClient.builder()
                .serverUrl(Constants.URL_GRAPHQL)
                .build();

        apolloClient.mutate(new LoginViaEmailMutation(email, password, deviceName))
                .enqueue(new ApolloCall.Callback<LoginViaEmailMutation.Data>() {
                    @Override
                    public void onResponse(@NonNull Response<LoginViaEmailMutation.Data> response) {
                        activity.runOnUiThread(() -> {
                            hideLoadingAnimation();

                            if (!response.hasErrors()) {
                                Log.i("loginViaEmail", "onResponse: " + response);

                                SharedPreferences.Editor editor = context
                                        .getSharedPreferences(Constants.PREFERENCE_LOGIN, Context.MODE_PRIVATE).edit();
                                editor.putString(Constants.PREFERENCE_USER_TOKEN_KEY, response.getData().loginViaEmail().token()).apply();
                                editor.putBoolean(Constants.PREFERENCE_IS_LOGGED_IN_KEY, true).apply();

                                startActivity(new Intent(context, MainActivity.class));
                                activity.finish();
                            } else {
                                Log.e("loginViaEmail", "Errors: " + response.getErrors());
                                if (Arrays.toString(response.getErrors().get(0).getCustomAttributes().values().toArray()).contains("code=401")) {
                                    Snackbar snackbar = Snackbar.make(view,
                                            context.getResources().getString(R.string.login_error_credentials),
                                            BaseTransientBottomBar.LENGTH_LONG)
                                            .setBackgroundTint(context.getResources().getColor(R.color.red))
                                            .setTextColor(context.getResources().getColor(R.color.white))
                                            .setAnimationMode(BaseTransientBottomBar.ANIMATION_MODE_SLIDE);
                                    TextView snackText = (snackbar.getView()).findViewById(R.id.snackbar_text);
                                    snackText.setTypeface(ResourcesCompat.getFont(context, R.font.nunito_semi_bold));
                                    snackbar.show();
                                }
                            }
                        });
                    }

                    @Override
                    public void onFailure(@NonNull ApolloException e) {
                        Log.e("loginViaEmail", "onFailure: " + e.getMessage());
                        activity.runOnUiThread(() -> {
                            Snackbar snackbar = Snackbar.make(view,
                                    context.getResources().getString(R.string.network_error_failure),
                                    BaseTransientBottomBar.LENGTH_LONG)
                                    .setBackgroundTint(context.getResources().getColor(R.color.red))
                                    .setTextColor(context.getResources().getColor(R.color.white))
                                    .setAnimationMode(BaseTransientBottomBar.ANIMATION_MODE_SLIDE);
                            TextView snackText = (snackbar.getView()).findViewById(R.id.snackbar_text);
                            snackText.setTypeface(ResourcesCompat.getFont(context, R.font.nunito_semi_bold));
                            snackbar.show();
                        });
                    }
                });
    }

    private void showLoadingAnimation() {
        loginBinding.btnLoginLogin.setText(null);
        loginBinding.btnLoginLogin.setClickable(false);
        loginBinding.cpiLogin.setVisibility(View.VISIBLE);
    }

    private void hideLoadingAnimation() {
        loginBinding.cpiLogin.setVisibility(View.GONE);
        loginBinding.btnLoginLogin.setClickable(true);
        loginBinding.btnLoginLogin.setText(context.getResources().getString(R.string.login_btn_login));
    }

    private boolean isEmailValid(String email) {
        return Patterns.EMAIL_ADDRESS.matcher(email).matches();
    }
}

File: app/src/main/java/com/xeniac/warrantyroster_manager/landingactivity/loginfragment/LoginFragment.kt
package com.xeniac.warrantyroster_manager.landingactivity.loginfragment

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.util.Patterns
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.navigation.NavController
import androidx.navigation.Navigation
import com.google.android.material.snackbar.BaseTransientBottomBar.LENGTH_INDEFINITE
import com.google.android.material.snackbar.BaseTransientBottomBar.LENGTH_LONG
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import com.xeniac.warrantyroster_manager.Constants
import com.xeniac.warrantyroster_manager.NetworkHelper
import com.xeniac.warrantyroster_manager.R
import com.xeniac.warrantyroster_manager.databinding.FragmentLoginBinding
import com.xeniac.warrantyroster_manager.mainactivity.MainActivity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext

class LoginFragment : Fragment(R.layout.fragment_login) {

    private var _binding: FragmentLoginBinding? = null
    private val binding get() = _binding!!
    private lateinit var navController: NavController

    private lateinit var firebaseAuth: FirebaseAuth

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        _binding = FragmentLoginBinding.bind(view)
        firebaseAuth = FirebaseAuth.getInstance()
        navController = Navigation.findNavController(view)

        textInputsBackgroundColor()
        textInputsStrokeColor()
        forgotPwOnClick()
        registerOnClick()
        loginOnClick()
        loginActionDone()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun textInputsBackgroundColor() {
        binding.tiLoginEditEmail.setOnFocusChangeListener { _, isFocused ->
            if (isFocused) {
                binding.tiLoginLayoutEmail.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.background)
            } else {
                binding.tiLoginLayoutEmail.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.grayLight)
            }
        }

        binding.tiLoginEditPassword.setOnFocusChangeListener { _, isFocused ->
            if (isFocused) {
                binding.tiLoginLayoutPassword.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.background)
            } else {
                binding.tiLoginLayoutPassword.boxBackgroundColor =
                    ContextCompat.getColor(requireContext(), R.color.grayLight)
            }
        }
    }

    private fun textInputsStrokeColor() {
        binding.tiLoginEditEmail.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(
                inputEmail: CharSequence?, start: Int, before: Int, count: Int
            ) {
                binding.tiLoginLayoutEmail.isErrorEnabled = false
                binding.tiLoginLayoutEmail.boxStrokeColor =
                    ContextCompat.getColor(requireContext(), R.color.blue)
            }

            override fun afterTextChanged(s: Editable?) {
            }
        })

        binding.tiLoginEditPassword.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(
                inputEmail: CharSequence?, start: Int, before: Int, count: Int
            ) {
                binding.tiLoginLayoutPassword.boxStrokeColor =
                    ContextCompat.getColor(requireContext(), R.color.blue)
            }

            override fun afterTextChanged(s: Editable?) {
            }
        })
    }

    private fun forgotPwOnClick() {
        binding.btnLoginForgotPw.setOnClickListener {
            navController.navigate(R.id.action_loginFragment_to_forgotPasswordFragment)
        }
    }

    private fun registerOnClick() {
        binding.btnLoginRegister.setOnClickListener {
            navController.navigate(R.id.action_loginFragment_to_registerFragment)
        }
    }

    private fun loginOnClick() {
        binding.btnLoginLogin.setOnClickListener {
            loginViaEmail()
        }
    }

    private fun loginActionDone() {
        binding.tiLoginEditPassword.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                loginViaEmail()
            }
            false
        }
    }

    private fun loginViaEmail() {
        val inputMethodManager = requireContext()
            .getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        inputMethodManager.hideSoftInputFromWindow(binding.root.applicationWindowToken, 0)

        if (NetworkHelper.hasNetworkAccess(requireContext())) {
            getLoginInputs()
        } else {
            hideLoadingAnimation()
            Snackbar.make(
                binding.root,
                requireContext().getString(R.string.network_error_connection),
                LENGTH_INDEFINITE
            ).apply {
                setAction(requireContext().getString(R.string.network_error_retry)) { getLoginInputs() }
                show()
            }
        }
    }

    private fun getLoginInputs() {
        val email = binding.tiLoginEditEmail.text.toString().trim().lowercase()
        val password = binding.tiLoginEditPassword.text.toString().trim()

        if (email.isBlank()) {
            binding.tiLoginLayoutEmail.requestFocus()
            binding.tiLoginLayoutEmail.boxStrokeColor =
                ContextCompat.getColor(requireContext(), R.color.red)
        } else if (password.isBlank()) {
            binding.tiLoginLayoutPassword.requestFocus()
            binding.tiLoginLayoutPassword.boxStrokeColor =
                ContextCompat.getColor(requireContext(), R.color.red)
        } else {
            if (!isEmailValid(email)) {
                binding.tiLoginLayoutEmail.requestFocus()
                binding.tiLoginLayoutEmail.error =
                    requireContext().getString(R.string.login_error_email)
            } else {
                loginViaEmailAuth(email, password)
            }
        }
    }

    private fun loginViaEmailAuth(email: String, password: String) {
        showLoadingAnimation()
        CoroutineScope(Dispatchers.IO).launch {
            try {
                firebaseAuth.signInWithEmailAndPassword(email, password).await().apply {
                    user?.let {
                        Log.i("loginViaEmail", "${it.email} logged in successfully.")

                        requireContext().getSharedPreferences(
                            Constants.PREFERENCE_LOGIN, Context.MODE_PRIVATE
                        ).edit().apply {
                            putBoolean(Constants.PREFERENCE_IS_LOGGED_IN_KEY, true)
                            apply()
                        }

                        withContext(Dispatchers.Main) {
                            hideLoadingAnimation()
                            Intent(requireContext(), MainActivity::class.java).apply {
                                startActivity(this)
                                requireActivity().finish()
                            }
                        }
                    }
                }
            } catch (e: Exception) {
                Log.e("loginViaEmail", "Exception: ${e.message}")
                withContext(Dispatchers.Main) {
                    hideLoadingAnimation()
                    if (e.toString().contains(
                            "The password is invalid or the user does not have a password"
                        )
                    ) {
                        Snackbar.make(
                            binding.root,
                            requireContext().getString(R.string.login_error_credentials),
                            LENGTH_LONG
                        ).show()
                    } else {
                        Snackbar.make(
                            binding.root,
                            requireContext().getString(R.string.network_error_failure),
                            LENGTH_LONG
                        ).show()
                    }
                }
            }
        }
    }

    private fun showLoadingAnimation() {
        binding.tiLoginEditEmail.isEnabled = false
        binding.tiLoginEditPassword.isEnabled = false
        binding.btnLoginLogin.isClickable = false
        binding.btnLoginLogin.text = null
        binding.cpiLogin.visibility = VISIBLE
    }

    private fun hideLoadingAnimation() {
        binding.tiLoginEditEmail.isEnabled = true
        binding.tiLoginEditPassword.isEnabled = true
        binding.btnLoginLogin.isClickable = true
        binding.btnLoginLogin.text =
            requireContext().getString(R.string.login_btn_login)
        binding.cpiLogin.visibility = GONE
    }

    private fun isEmailValid(email: String): Boolean {
        return Patterns.EMAIL_ADDRESS.matcher(email).matches()
    }
}